### PR TITLE
[OpenMP][Clang] Migrate OpenMP UserDefinedMapper from Clang to OMPIRBuilder

### PR DESCRIPTION
**Pull request by TIFitis**
This patch migrates the OpenMP UserDefinedMapper codegen from Clang to the OpenMPIRBuilder. I will be adding further patches in the near future so that OpenMP dialect in MLIR can make use of these.

### CODE REVIEW

**Comment by jplehr** [comment]
Thank you. Will take a closer look next week.
So far, I ran this through one of our buildbot configs and did not see an issue there.

**Review by jplehr — COMMENTED** [review]
There is a nit, but I'm also not very familiar with this piece of code.

**Comment by TIFitis** [comment]
@jdoerfert @jsjodin Polite request for review :)

File: llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h [hunk 1]
```diff
@@ -2801,6 +2801,67 @@ class OpenMPIRBuilder {
   using GenMapInfoCallbackTy =
       function_ref<MapInfosTy &(InsertPointTy CodeGenIP)>;
 
+private:
+  /// Emit the array initialization or deletion portion for user-defined mapper
+  /// code generation. First, it evaluates whether an array section is mapped
+  /// and whether the \a MapType instructs to delete this section. If \a IsInit
+  /// is true, and \a MapType indicates to not delete this array, array
+  /// initialization code is generated. If \a IsInit is false, and \a MapType
+  /// indicates to not this array, array deletion code is generated.
```
Comment (dhruvachak) [inline llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:2810]: typo: indicates to not this array
please rephrase ^^

File: llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h [hunk 2]
```diff
@@ -2801,6 +2801,67 @@ class OpenMPIRBuilder {
...
+  void emitUDMapperArrayInitOrDel(Function *MapperFn, llvm::Value *Handle,
```
Comment (dhruvachak) [inline llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:2811]: Document Handle, e.g. add a \param for it.

**Comment by TIFitis** [comment]
@jdoerfert @jsjodin Polite request for review :)

File: llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h (hunk 2 above)
Comment (TIFitis) [reply llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:2811]: I've updated the name to MapperHandle to better describe the param. Do we need a /param description for it given that this is a private member function?

File: clang/lib/CodeGen/CGOpenMPRuntime.cpp [hunk 3]
```diff
@@ -9041,257 +9041,65 @@ void CGOpenMPRuntime::emitUserDefinedMapper(const OMPDeclareMapperDecl *D,
     return;
   ASTContext &C = CGM.getContext();
   QualType Ty = D->getType();
-  QualType PtrTy = C.getPointerType(Ty).withRestrict();
-  QualType Int64Ty = C.getIntTypeForBitwidth(/*DestWidth=*/64, /*Signed=*/true);
   auto *MapperVarDecl =
       cast<VarDecl>(cast<DeclRefExpr>(D->getMapperVarRef())->getDecl());
-  SourceLocation Loc = D->getLocation();
```
Comment (jsjodin) [inline clang/lib/CodeGen/CGOpenMPRuntime.cpp:9048]: Is the location information propagated? I am wondering if debug info might be lost like what happened in the reduction migration patch.

Comment (TIFitis) [reply clang/lib/CodeGen/CGOpenMPRuntime.cpp:9048]: This `SourceLocation` is defined but never used in the original Clang code itself.

**Review by jsjodin — APPROVED** [review]
LGTM

**Pull request merged**